aten/src/ATen/core/VariableHooksInterface.h

```diff
@@ -40,23 +40,25 @@ namespace impl {
 
 struct TORCH_API VariableHooksInterface {
   virtual ~VariableHooksInterface() = default;
-  virtual Tensor tensor_data(const Tensor&) const = 0;
-  virtual Tensor variable_data(const Tensor&) const = 0;
-  virtual const std::shared_ptr<torch::autograd::Node>& grad_fn(const Tensor&) const = 0;
-  virtual unsigned _register_hook(const Tensor&, std::function<Tensor(const Tensor&)> hook) const = 0;
-  virtual void remove_hook(const Tensor&, unsigned pos) const = 0;
-  virtual bool is_view(const Tensor&) const = 0;
-  virtual const Tensor& base(const Tensor&) const = 0;
-  virtual const std::string& name(const Tensor&) const = 0;
-  virtual bool is_leaf(const Tensor&) const = 0;
-  virtual int64_t output_nr(const Tensor&) const = 0;
-  virtual void set_data(const Tensor&, const Tensor&) const = 0;
-  virtual Tensor data(const Tensor&) const = 0;
-  virtual int64_t _version(const Tensor&) const = 0;
-  virtual void retain_grad(const Tensor&) const = 0;
-  virtual bool retains_grad(const Tensor&) const = 0;
+  virtual TensorBase tensor_data(const TensorBase&) const = 0;
+  virtual TensorBase variable_data(const TensorBase&) const = 0;
+  virtual const std::shared_ptr<torch::autograd::Node>& grad_fn(const TensorBase&) const = 0;
+  virtual unsigned _register_hook(
+      const TensorBase&,
+      std::function<TensorBase(const TensorBase&)> hook) const = 0;
+  virtual void remove_hook(const TensorBase&, unsigned pos) const = 0;
+  virtual bool is_view(const TensorBase&) const = 0;
+  virtual const TensorBase& base(const TensorBase&) const = 0;
+  virtual const std::string& name(const TensorBase&) const = 0;
+  virtual bool is_leaf(const TensorBase&) const = 0;
+  virtual int64_t output_nr(const TensorBase&) const = 0;
+  virtual void set_data(const TensorBase&, const TensorBase&) const = 0;
+  virtual TensorBase data(const TensorBase&) const = 0;
+  virtual int64_t _version(const TensorBase&) const = 0;
+  virtual void retain_grad(const TensorBase&) const = 0;
+  virtual bool retains_grad(const TensorBase&) const = 0;
   virtual void _backward(const Tensor&, TensorList, const c10::optional<Tensor>&, c10::optional<bool>, bool) const = 0;
```

Comment:
This is an example where invariance of containers means you can't replace Tensor with TensorBase. Is there a downside to not converting the signature here?

Reply:
I might look into it if it becomes relevant, but at the moment I'm focused on operator kernels. Kernels both take the largest proportion of the compile time, and also don't really use autograd so it isn't relevant here.

The above comment about needing to reallocate to convert between `std::vector<Tensor>` and `std::vector<TensorBase>` is also an important factor here.

Reply:
That's fine. I was just a little surprised that it was OK not to update this haha.

Reply:
This works out because `VariableHooksInterface.h` is only included from the `Tensor.cpp` file, not any headers. So, it's okay to include `Tensor.h`.